src/rgw/cls_fifo_legacy.cc

```diff
@@ -1004,6 +1012,10 @@ struct NewHeadPreparer : public Completion<NewHeadPreparer> {
       lderr(f->cct) << __PRETTY_FUNCTION__ << ":" << __LINE__
 		    << " _prepare_new_part failed: r=" << r
 		    << " tid=" << tid << dendl;
+      std::unique_lock l(f->m);
```

Comment:
Can we override Completion::complete in NewHeadPreparer and do the notification there and call the parent's (Completion) complete? So that we don't need to repeat this code block several times. before each complete() statement.